pkg/dockerregistry/server/writelimiter/semaphore.go

```diff
@@ -0,0 +1,24 @@
+package writelimiter
+
+type semaphore struct {
+	c chan struct{}
+}
+
+func newSemaphore(count int) *semaphore {
+	return &semaphore{
+		c: make(chan struct{}, count),
+	}
+}
+
+func (s *semaphore) Up() {
+	<-s.c
+}
+
+func (s *semaphore) TryDown() bool {
```

Comment:
Re context: https://golang.org/pkg/context/#WithTimeout is what we should be using